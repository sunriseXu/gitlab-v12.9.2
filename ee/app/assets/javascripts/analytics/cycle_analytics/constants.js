import { __ } from '~/locale';

export const PROJECTS_PER_PAGE = 50;

export const DEFAULT_DAYS_IN_PAST = 30;

export const EVENTS_LIST_ITEM_LIMIT = 50;

export const EMPTY_STAGE_TEXT = {
  issue: __(
    'The issue stage shows the time it takes from creating an issue to assigning the issue to a milestone, or add the issue to a list on your Issue Board. Begin creating issues to see data for this stage.',
  ),
  plan: __(
    'The planning stage shows the time from the previous step to pushing your first commit. This time will be added automatically once you push your first commit.',
  ),
  code: __(
    'The coding stage shows the time from the first commit to creating the merge request. The data will automatically be added here once you create your first merge request.',
  ),
  test: __(
    'The testing stage shows the time GitLab CI takes to run every pipeline for the related merge request. The data will automatically be added after your first pipeline finishes running.',
  ),
  review: __(
    'The review stage shows the time from creating the merge request to merging it. The data will automatically be added after you merge your first merge request.',
  ),
  staging: __(
    'The staging stage shows the time between merging the MR and deploying code to the production environment. The data will be automatically added once you deploy to production for the first time.',
  ),
  production: __(
    'The production stage shows the total time it takes between creating an issue and deploying the code to production. The data will be automatically added once you have completed the full idea to production cycle.',
  ),
};

export const DEFAULT_STAGE_NAMES = [...Object.keys(EMPTY_STAGE_TEXT), 'total'];

export const TASKS_BY_TYPE_SUBJECT_ISSUE = 'Issue';
export const TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST = 'MergeRequest';
export const TASKS_BY_TYPE_MAX_LABELS = 15;

export const TASKS_BY_TYPE_SUBJECT_FILTER_OPTIONS = {
  [TASKS_BY_TYPE_SUBJECT_ISSUE]: __('Issues'),
  [TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST]: __('Merge Requests'),
};

export const TASKS_BY_TYPE_FILTERS = {
  SUBJECT: 'SUBJECT',
  LABEL: 'LABEL',
};

export const STAGE_ACTIONS = {
  SELECT: 'selectStage',
  EDIT: 'editStage',
  REMOVE: 'removeStage',
  HIDE: 'hideStage',
  CREATE: 'createStage',
  UPDATE: 'updateStage',
};

export const STAGE_NAME = {
  TOTAL: 'total',
  PRODUCTION: 'production',
};
