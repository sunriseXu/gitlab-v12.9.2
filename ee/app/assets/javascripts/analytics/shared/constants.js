export const dateFormats = {
  isoDate: 'yyyy-mm-dd',
  defaultDate: 'mmm d, yyyy',
  defaultDateTime: 'mmm d, yyyy h:MMtt',
};

export const scatterChartLineProps = {
  default: {
    type: 'line',
    showSymbol: false,
    // By default zlevel is 2 for all series types.
    // By increasing the zlevel to 3 we make sure that the trendline gets drawn in front of the dots in the chart.
    zlevel: 3,
  },
};

export const LAST_ACTIVITY_AT = 'last_activity_at';

export const DATE_RANGE_LIMIT = 180;

export const OFFSET_DATE_BY_ONE = 1;
