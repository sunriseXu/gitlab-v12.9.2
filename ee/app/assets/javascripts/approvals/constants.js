import { __ } from '~/locale';

export const TYPE_USER = 'user';
export const TYPE_GROUP = 'group';
export const TYPE_HIDDEN_GROUPS = 'hidden_groups';

export const RULE_TYPE_FALLBACK = 'fallback';
export const RULE_TYPE_REGULAR = 'regular';
export const RULE_TYPE_REPORT_APPROVER = 'report_approver';
export const RULE_TYPE_CODE_OWNER = 'code_owner';
export const RULE_TYPE_ANY_APPROVER = 'any_approver';
export const RULE_NAME_ANY_APPROVER = 'All Members';

export const VULNERABILITY_CHECK_NAME = 'Vulnerability-Check';
export const LICENSE_CHECK_NAME = 'License-Check';

export const APPROVAL_RULE_CONFIGS = {
  [VULNERABILITY_CHECK_NAME]: {
    title: __('Vulnerability-Check'),
    popoverText: __(
      'A merge request approval is required when a security report contains a new vulnerability of high, critical, or unknown severity.',
    ),
    documentationText: __('Learn more about Vulnerability-Check'),
  },
  [LICENSE_CHECK_NAME]: {
    title: __('License-Check'),
    popoverText: __(
      'A merge request approval is required when the license compliance report contains a blacklisted license.',
    ),
    documentationText: __('Learn more about License-Check'),
  },
};
