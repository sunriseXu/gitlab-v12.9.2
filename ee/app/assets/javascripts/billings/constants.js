export const TABLE_TYPE_DEFAULT = 'default';
export const TABLE_TYPE_FREE = 'free';
export const TABLE_TYPE_TRIAL = 'trial';
