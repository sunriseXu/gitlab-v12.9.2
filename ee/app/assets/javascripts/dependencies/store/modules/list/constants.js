import { __, s__ } from '~/locale';

export const SORT_FIELDS = {
  name: s__('Dependencies|Component name'),
  packager: s__('Dependencies|Packager'),
  severity: s__('Vulnerability|Severity'),
};

export const SORT_ORDER = {
  ascending: 'asc',
  descending: 'desc',
};

export const REPORT_STATUS = {
  ok: 'ok',
  jobNotSetUp: 'job_not_set_up',
  jobFailed: 'job_failed',
  noDependencies: 'no_dependencies',
  incomplete: 'no_dependency_files',
};

export const FILTER = {
  all: 'all',
  vulnerable: 'vulnerable',
};

export const FETCH_ERROR_MESSAGE = __(
  'Error fetching the dependency list. Please check your network connection and try again.',
);
