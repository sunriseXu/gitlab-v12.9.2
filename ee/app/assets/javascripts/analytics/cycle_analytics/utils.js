import { isNumber } from 'underscore';
import dateFormat from 'dateformat';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { convertToSnakeCase } from '~/lib/utils/text_utility';
import { hideFlash } from '~/flash';
import {
  newDate,
  dayAfter,
  secondsToDays,
  getDatesInRange,
  getDayDifference,
  getDateInPast,
  getDateInFuture,
} from '~/lib/utils/datetime_utility';
import { dateFormats } from '../shared/constants';
import { STAGE_NAME } from './constants';
import { toYmd } from '../shared/utils';

const EVENT_TYPE_LABEL = 'label';

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

export const toggleSelectedLabel = ({ selectedLabelIds = [], value = null }) => {
  if (!value) return selectedLabelIds;
  return selectedLabelIds.includes(value)
    ? selectedLabelIds.filter(v => v !== value)
    : [...selectedLabelIds, value];
};

export const isStartEvent = ev => Boolean(ev) && Boolean(ev.canBeStartEvent) && ev.canBeStartEvent;

export const eventToOption = (obj = null) => {
  if (!obj || (!obj.text && !obj.identifier)) return null;
  const { name: text = '', identifier: value = null } = obj;
  return { text, value };
};

export const getAllowedEndEvents = (events = [], targetIdentifier = null) => {
  if (!targetIdentifier || !events.length) return [];
  const st = events.find(({ identifier }) => identifier === targetIdentifier);
  return st && st.allowedEndEvents ? st.allowedEndEvents : [];
};

export const eventsByIdentifier = (events = [], targetIdentifier = []) => {
  if (!targetIdentifier || !targetIdentifier.length || !events.length) return [];
  return events.filter(({ identifier = '' }) => targetIdentifier.includes(identifier));
};

export const isLabelEvent = (labelEvents = [], ev = null) =>
  Boolean(ev) && labelEvents.length && labelEvents.includes(ev);

export const getLabelEventsIdentifiers = (events = []) =>
  events.filter(ev => ev.type && ev.type === EVENT_TYPE_LABEL).map(i => i.identifier);

/**
 * Checks if the specified stage is in memory or persisted to storage based on the id
 *
 * Default value stream analytics stages are initially stored in memory, when they are first
 * created the id for the stage is the name of the stage in lowercase. This string id
 * is used to fetch stage data (events, median calculation)
 *
 * When either a custom stage is created or an edit is made to a default stage then the
 * default stages get persisted to storage and will have a numeric id. The new numeric
 * id should then be used to access stage data
 *
 */
export const isPersistedStage = ({ custom, id }) => custom || isNumber(id);

/**
 * Returns the the correct slug to use for a stage
 * default stages use the snakecased title of the stage, while custom
 * stages will have a numeric id
 *
 * @param {Object} obj
 * @param {string} obj.title - title of the stage
 * @param {number} obj.id - numerical object id available for custom stages
 * @param {boolean} obj.custom - boolean flag indicating a custom stage
 * @returns {(number|string)} Returns a numerical id for customs stages and string for default stages
 */
const stageUrlSlug = ({ id, title, custom = false }) => {
  if (custom) return id;
  // We still use 'production' as the id to access this stage, even though the title is 'Total'
  return title.toLowerCase() === STAGE_NAME.TOTAL
    ? STAGE_NAME.PRODUCTION
    : convertToSnakeCase(title);
};

export const transformRawStages = (stages = []) =>
  stages.map(({ id, title, name = '', custom = false, ...rest }) => ({
    ...convertObjectPropsToCamelCase(rest, { deep: true }),
    id,
    title,
    custom,
    slug: isPersistedStage({ custom, id }) ? id : stageUrlSlug({ custom, id, title }),
    // the name field is used to create a stage, but the get request returns title
    name: name.length ? name : title,
  }));

export const transformRawTasksByTypeData = (data = []) => {
  if (!data.length) return [];
  return data.map(d => convertObjectPropsToCamelCase(d, { deep: true }));
};

/**
 * Takes the duration data for selected stages, transforms the date values and returns
 * the data in a flattened array
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then transformed and flattened into the following format;
 * [
 *  {
 *    'duration_in_seconds': 1234,
 *    'finished_at': '2019-09-02'
 *  },
 *  ...
 * ]
 *
 * @param {Array} data - The duration data for selected stages
 * @returns {Array} An array with each item being an object containing the duration_in_seconds and finished_at values for an event
 */
export const flattenDurationChartData = data =>
  data
    .map(stage =>
      stage.data.map(event => {
        const date = new Date(event.finished_at);
        return {
          ...event,
          finished_at: dateFormat(date, dateFormats.isoDate),
        };
      }),
    )
    .flat();

/**
 * Takes the duration data for selected stages, groups the data by day and calculates the total duration
 * per day.
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then computed and transformed into a format that can be passed to the chart:
 * [
 *  ['2019-09-02', 7, '2019-09-02'],
 *  ['2019-09-03', 10, '2019-09-03'],
 *  ['2019-09-04', 8, '2019-09-04'],
 *  ...
 * ]
 *
 * In the data above, each array i represents a point in the scatterplot with the following data:
 * i[0] = date, displayed on x axis
 * i[1] = metric, displayed on y axis
 * i[2] = date, used in the tooltip
 *
 * @param {Array} data - The duration data for selected stages
 * @param {Date} startDate - The globally selected cycle analytics start date
 * @param {Date} endDate - The globally selected cycle analytics end date
 * @returns {Array} An array with each item being another arry of three items (plottable date, computed total, tooltip display date)
 */
export const getDurationChartData = (data, startDate, endDate) => {
  const flattenedData = flattenDurationChartData(data);
  const eventData = [];

  for (
    let currentDate = newDate(startDate);
    currentDate <= endDate;
    currentDate = dayAfter(currentDate)
  ) {
    const currentISODate = dateFormat(newDate(currentDate), dateFormats.isoDate);
    const valuesForDay = flattenedData.filter(object => object.finished_at === currentISODate);
    const summedData = valuesForDay.reduce((total, value) => total + value.duration_in_seconds, 0);
    const summedDataInDays = secondsToDays(summedData);

    if (summedDataInDays) eventData.push([currentISODate, summedDataInDays, currentISODate]);
  }

  return eventData;
};

/**
 * Takes the offset duration data for selected stages and calls getDurationChartData to compute the totals.
 * The data is then transformed into a format expected by the scatterplot;
 *
 * [
 *   ['2019-09-02', 7],
 *   ...
 * ]
 *
 * The transformation works by calling getDateInPast on the provided startDate and endDate in order to match
 * the startDate and endDate fetched when making the API call to fetch the data.
 *
 * In order to map the offset data to plottable points within the chart's range, getDateInFuture is called
 * on the data series with the same offest used for getDateInPast. This creates plottable data that matches up
 * with the data being displayed on the chart.
 *
 * @param {Array} data - The computed, plottable duration chart data
 * @param {Date} startDate - The globally selected cycle analytics start date
 * @param {Date} endDate - The globally selected cycle analytics end date
 * @returns {Array} An array with each item being another arry of two items (date, computed total)
 */
export const getDurationChartMedianData = (data, startDate, endDate) => {
  const offsetValue = getDayDifference(startDate, endDate);
  const offsetEndDate = getDateInPast(endDate, offsetValue);
  const offsetStartDate = getDateInPast(startDate, offsetValue);

  const offsetDurationData = getDurationChartData(data, offsetStartDate, offsetEndDate);

  const result = offsetDurationData.map(event => [
    dateFormat(getDateInFuture(new Date(event[0]), offsetValue), dateFormats.isoDate),
    event[1],
  ]);

  return result;
};

export const orderByDate = (a, b, dateFmt = datetime => new Date(datetime).getTime()) =>
  dateFmt(a) - dateFmt(b);

/**
 * Takes a dictionary of dates and the associated value, sorts them and returns just the value
 *
 * @param {Object.<Date, number>} series - Key value pair of dates and the value for that date
 * @returns {number[]} The values of each key value pair
 */
export const flattenTaskByTypeSeries = (series = {}) =>
  Object.entries(series)
    .sort((a, b) => orderByDate(a[0], b[0]))
    .map(dataSet => dataSet[1]);

/**
 * @typedef {Object} RawTasksByTypeData
 * @property {Object} label - Raw data for a group label
 * @property {Array} series - Array of arrays with date and associated value ie [ ['2020-01-01', 10],['2020-01-02', 10] ]

 * @typedef {Object} TransformedTasksByTypeData
 * @property {Array} groupBy - The list of dates for the range of data in each data series
 * @property {Array} data - An array of the data values for each series
 * @property {Array} seriesNames - Names of the series to be charted ie label names
 */

/**
 * Takes the raw tasks by type data and generates an array of data points,
 * an array of data series and an array of data labels for the given time period.
 *
 * Currently the data is transformed to support use in a stacked column chart:
 * https://gitlab-org.gitlab.io/gitlab-ui/?path=/story/charts-stacked-column-chart--stacked
 *
 * @param {Object} obj
 * @param {RawTasksByTypeData[]} obj.data - array of raw data, each element contains a label and series
 * @param {Date} obj.startDate - start date in ISO date format
 * @param {Date} obj.endDate - end date in ISO date format
 *
 * @returns {TransformedTasksByTypeData} The transformed data ready for use in charts
 */
export const getTasksByTypeData = ({ data = [], startDate = null, endDate = null }) => {
  if (!startDate || !endDate || !data.length) {
    return {
      groupBy: [],
      data: [],
      seriesNames: [],
    };
  }

  const groupBy = getDatesInRange(startDate, endDate, toYmd).sort(orderByDate);
  const zeroValuesForEachDataPoint = groupBy.reduce(
    (acc, date) => ({
      ...acc,
      [date]: 0,
    }),
    {},
  );

  const transformed = data.reduce(
    (acc, curr) => {
      const {
        label: { title },
        series,
      } = curr;
      acc.seriesNames = [...acc.seriesNames, title];
      acc.data = [
        ...acc.data,
        // adds 0 values for each data point and overrides with data from the series
        flattenTaskByTypeSeries({ ...zeroValuesForEachDataPoint, ...Object.fromEntries(series) }),
      ];
      return acc;
    },
    {
      data: [],
      seriesNames: [],
    },
  );

  return {
    ...transformed,
    groupBy,
  };
};
