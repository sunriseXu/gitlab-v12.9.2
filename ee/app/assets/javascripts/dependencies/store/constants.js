import { s__ } from '~/locale';
import { FILTER } from './modules/list/constants';

// eslint-disable-next-line import/prefer-default-export
export const DEPENDENCY_LIST_TYPES = {
  all: {
    namespace: 'allDependencies',
    label: s__('Dependencies|All'),
    initialState: {
      filter: FILTER.all,
    },
  },
  vulnerable: {
    namespace: 'vulnerableDependencies',
    label: s__('Dependencies|Vulnerable components'),
    initialState: {
      filter: FILTER.vulnerable,
    },
  },
};
