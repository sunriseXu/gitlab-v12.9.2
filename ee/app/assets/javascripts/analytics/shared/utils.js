import dateFormat from 'dateformat';
import { dateFormats } from './constants';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';

export const toYmd = date => dateFormat(date, dateFormats.isoDate);

export default {
  toYmd,
};

export const formattedDate = d => dateFormat(d, dateFormats.defaultDate);

/**
 * Creates a group object from a dataset. Returns null if no groupId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A group object
 */
export const buildGroupFromDataset = dataset => {
  const { groupId, groupName, groupFullPath, groupAvatarUrl } = dataset;

  if (groupId) {
    return {
      id: Number(groupId),
      name: groupName,
      full_path: groupFullPath,
      avatar_url: groupAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates a project object from a dataset. Returns null if no projectId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A project object
 */
export const buildProjectFromDataset = dataset => {
  const { projectId, projectName, projectPathWithNamespace, projectAvatarUrl } = dataset;

  if (projectId) {
    return {
      id: Number(projectId),
      name: projectName,
      path_with_namespace: projectPathWithNamespace,
      avatar_url: projectAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates an array of project objects from a json string. Returns null if no projects are present.
 *
 * @param {String} data - JSON encoded array of projects
 * @returns {Array} - An array of project objects
 */
const buildProjectsFromJSON = (projects = []) => {
  if (!projects.length) return [];
  return JSON.parse(projects);
};

/**
 * Builds the initial data object for cycle analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  groupId = null,
  createdBefore = null,
  createdAfter = null,
  projects = null,
  groupName = null,
  groupFullPath = null,
  groupAvatarUrl = null,
} = {}) => ({
  group: groupId
    ? convertObjectPropsToCamelCase(
        buildGroupFromDataset({ groupId, groupName, groupFullPath, groupAvatarUrl }),
      )
    : null,
  createdBefore: createdBefore ? new Date(createdBefore) : null,
  createdAfter: createdAfter ? new Date(createdAfter) : null,
  selectedProjects: projects
    ? buildProjectsFromJSON(projects).map(convertObjectPropsToCamelCase)
    : [],
});
