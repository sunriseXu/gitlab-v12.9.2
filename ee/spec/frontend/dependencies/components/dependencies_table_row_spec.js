import { shallowMount } from '@vue/test-utils';
import DependenciesTableRow from 'ee/dependencies/components/dependencies_table_row.vue';
import DependencyVulnerability from 'ee/dependencies/components/dependency_vulnerability.vue';
import { MAX_DISPLAYED_VULNERABILITIES_PER_DEPENDENCY } from 'ee/dependencies/components/constants';
import { makeDependency } from './utils';
import mockDataVulnerabilities from '../../security_dashboard/store/vulnerabilities/data/mock_data_vulnerabilities';

describe('DependenciesTableRow component', () => {
  let wrapper;

  const factory = ({ propsData, ...options } = {}) => {
    wrapper = shallowMount(DependenciesTableRow, {
      ...options,
      propsData: { ...propsData },
    });
  };

  const findVulnerabilities = () => wrapper.findAll(DependencyVulnerability).wrappers;
  const findExcessMessage = () => wrapper.find('.js-excess-message');
  const expectVulnerabilitiesCollapsed = () => expect(findVulnerabilities()).toHaveLength(0);

  const expectVulnerabilitiesExpanded = vulnerabilities => {
    const wrappers = findVulnerabilities();
    expect(wrappers).toHaveLength(vulnerabilities.length);
    wrappers.forEach((vulnerabilityWrapper, i) => {
      expect(vulnerabilityWrapper.isVisible()).toBe(true);
      expect(vulnerabilityWrapper.props().vulnerability).toEqual(vulnerabilities[i]);
    });
    expect(findExcessMessage().exists()).toBe(false);
  };

  const clickToggle = () => {
    wrapper.find('.js-vulnerabilities-toggle').vm.$emit('click');
    return wrapper.vm.$nextTick();
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when passed no props', () => {
    beforeEach(() => {
      factory();
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when loading', () => {
    beforeEach(() => {
      factory({
        propsData: {
          isLoading: true,
        },
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when a dependency with no vulnerabilities is loaded', () => {
    beforeEach(() => {
      factory({
        propsData: {
          isLoading: false,
          dependency: makeDependency({ vulnerabilities: [] }),
        },
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when a dependency with vulnerabilities is loaded', () => {
    beforeEach(() => {
      factory({
        propsData: {
          isLoading: false,
          dependency: makeDependency({ vulnerabilities: mockDataVulnerabilities }),
        },
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('when the list of vulnerabilities is expanded', () => {
      beforeEach(clickToggle);

      it('renders each vulnerability', () => {
        expectVulnerabilitiesExpanded(mockDataVulnerabilities);
      });

      describe('when clicking the toggle ', () => {
        beforeEach(clickToggle);

        it('closes the list of vulnerabilities', expectVulnerabilitiesCollapsed);
      });

      describe('when the dependency prop changes', () => {
        beforeEach(() => {
          wrapper.setProps({
            dependency: makeDependency({ vulnerabilities: mockDataVulnerabilities }),
          });

          return wrapper.vm.$nextTick();
        });

        it('closes the list of vulnerabilities', expectVulnerabilitiesCollapsed);
      });

      describe('when the isLoading prop changes', () => {
        beforeEach(() => {
          wrapper.setProps({
            isLoading: true,
          });

          return wrapper.vm.$nextTick();
        });

        it('closes the list of vulnerabilities', expectVulnerabilitiesCollapsed);
      });
    });
  });

  describe('when a dependency with a huge number vulnerabilities is loaded and expanded', () => {
    beforeEach(() => {
      const hugeNumberOfVulnerabilities = Array(1 + MAX_DISPLAYED_VULNERABILITIES_PER_DEPENDENCY)
        .fill(null)
        .map((_, id) => ({ id }));

      factory({
        propsData: {
          isLoading: false,
          dependency: makeDependency({ vulnerabilities: hugeNumberOfVulnerabilities }),
        },
      });

      return clickToggle();
    });

    it('does not render all of them', () => {
      expect(findVulnerabilities().length).toBe(MAX_DISPLAYED_VULNERABILITIES_PER_DEPENDENCY);
      expect(findExcessMessage().isVisible()).toBe(true);
    });
  });
});
