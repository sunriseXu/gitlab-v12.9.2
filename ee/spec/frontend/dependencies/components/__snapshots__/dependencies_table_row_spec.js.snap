// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependenciesTableRow component when a dependency with no vulnerabilities is loaded matches the snapshot 1`] = `
<div
  class="gl-responsive-table-row flex-md-column align-items-md-stretch px-2"
>
  <div
    class="d-md-flex align-items-baseline"
  >
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Status
      </div>
       
      <div
        class="table-mobile-content"
      >
        <span
          class="text-success-500 text-1"
        >
          <icon-stub
            class="align-middle mr-1"
            name="check-circle"
            size="16"
          />
          Safe
        
        </span>
      </div>
    </div>
     
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        
        Component
      
      </div>
       
      <div
        class="table-mobile-content"
      >
        left-pad
      </div>
    </div>
     
    <div
      class="table-section section-10 pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Version
      </div>
       
      <div
        class="table-mobile-content"
      >
        0.0.3
      </div>
    </div>
     
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Packager
      </div>
       
      <div
        class="table-mobile-content"
      >
        JavaScript (yarn)
      </div>
    </div>
     
    <div
      class="table-section section-15 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Location
      </div>
       
      <div
        class="table-mobile-content"
      >
        <a
          href="/a-group/a-project/blob/da39a3ee5e6b4b0d3255bfef95601890afd80709/yarn.lock"
        >
          yarn.lock
        </a>
      </div>
    </div>
     
    <div
      class="table-section section-15 section-wrap"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        License
      </div>
       
      <div
        class="table-mobile-content"
      >
        <dependency-license-links-stub
          licenses=""
          title="left-pad"
        />
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`DependenciesTableRow component when a dependency with vulnerabilities is loaded matches the snapshot 1`] = `
<div
  class="gl-responsive-table-row flex-md-column align-items-md-stretch px-2"
>
  <div
    class="d-md-flex align-items-baseline"
  >
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Status
      </div>
       
      <div
        class="table-mobile-content"
      >
        <gl-button-stub
          class="bold text-warning-700 text-1 text-decoration-none js-vulnerabilities-toggle"
          size="md"
          variant="link"
        >
          <icon-stub
            class="align-top text-secondary-700 d-none d-md-inline"
            name="chevron-down"
            size="16"
          />
          
          8 vulnerabilities
        
        </gl-button-stub>
      </div>
    </div>
     
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        
        Component
      
      </div>
       
      <div
        class="table-mobile-content"
      >
        left-pad
      </div>
    </div>
     
    <div
      class="table-section section-10 pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Version
      </div>
       
      <div
        class="table-mobile-content"
      >
        0.0.3
      </div>
    </div>
     
    <div
      class="table-section section-20 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Packager
      </div>
       
      <div
        class="table-mobile-content"
      >
        JavaScript (yarn)
      </div>
    </div>
     
    <div
      class="table-section section-15 section-wrap pr-md-3"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        Location
      </div>
       
      <div
        class="table-mobile-content"
      >
        <a
          href="/a-group/a-project/blob/da39a3ee5e6b4b0d3255bfef95601890afd80709/yarn.lock"
        >
          yarn.lock
        </a>
      </div>
    </div>
     
    <div
      class="table-section section-15 section-wrap"
    >
      <div
        class="table-mobile-header"
        role="rowheader"
      >
        License
      </div>
       
      <div
        class="table-mobile-content"
      >
        <dependency-license-links-stub
          licenses=""
          title="left-pad"
        />
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`DependenciesTableRow component when loading matches the snapshot 1`] = `
<div
  class="gl-responsive-table-row flex-md-column align-items-md-stretch px-2"
>
  <gl-skeleton-loading-stub
    class="d-flex flex-column justify-content-center h-auto"
    lines="1"
  />
   
  <!---->
</div>
`;

exports[`DependenciesTableRow component when passed no props matches the snapshot 1`] = `
<div
  class="gl-responsive-table-row flex-md-column align-items-md-stretch px-2"
>
  <gl-skeleton-loading-stub
    class="d-flex flex-column justify-content-center h-auto"
    lines="1"
  />
   
  <!---->
</div>
`;
